tests/sysendian/main.c

```diff
@@ -0,0 +1,235 @@
+#include <stdint.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include "getopt.h"
+#include "perftest.h"
+#include "sysendian.h"
+#include "warnp.h"
+
+/* Performance tests. */
+static const size_t perfsizes[] = {32, 64}; /* Bits. */
+static const size_t num_perf = sizeof(perfsizes) / sizeof(perfsizes[0]);
+static const size_t nbytes_perftest = 400000000;	/* 400 MB */
+static const size_t nbytes_warmup = 20000000;		/* 200 MB */
+
+/* Print array as a series of 8-bit hex values. */
+static void
+print_arr(const uint8_t * arr, size_t n)
+{
+	size_t i;
+
+	for (i = 0; i < n; i++)
+		printf("%02x ", arr[i]);
+	printf("\n");
+}
+
+/*
+ * Generate a little-endian and big-endian representation of 0, 1, 2... (N-1).
+ */
+#define MAKE_LE_BE(le, be, N) do {			\
+	uint ## N ## _t le_full;			\
+	uint ## N ## _t be_full;			\
+	uint8_t * le_ptr;				\
+	uint8_t * be_ptr;				\
+	uint8_t n;					\
+	uint8_t i;					\
+							\
+	le_ptr = (uint8_t *)&le_full;			\
+	be_ptr = (uint8_t *)&be_full;			\
+	n = N/8;					\
+	for (i = 0; i < n; i++) {			\
+		le_ptr[i] = i;				\
+		be_ptr[i] = n - (i+1);			\
+	}						\
+	le = le_full;					\
+	be = be_full;					\
+} while (0)
+
+/**
+ * TEST_ENDIAN(ENDIAN, N, enc_pre, enc_post, dec_pre, dec_post):
+ * Check $N$-bit $ENDIAN$ (should be le or be) conversions.  The remaining
+ * variables must be uint$N$_t, with the $_pre$ values being the input to
+ * the ...enc() or ...dec() functions, and the $_post$ values being the
+ * desired output.
+ *
+ * Implementation note: Moving the values through le_v and be_v stops the
+ * compiler from generating the answers at compile-time.
+ */
+#define TEST_ENDIAN(ENDIAN, N, enc_pre, enc_post, dec_pre, dec_post) \
+do {							\
+	uint8_t arr[N/8];				\
+	uint ## N ## _t val;				\
+							\
+	ENDIAN ## N ## enc(arr, enc_pre);			\
+	if (memcmp(arr, &enc_post, N/8)) {		\
+		warnp("Failed " #ENDIAN #N "enc");	\
+		print_arr(arr, N/8);			\
+		goto err0;				\
+	}						\
+							\
+	val = ENDIAN ## N ## dec(&dec_pre);		\
+	if (val != dec_post) {			\
+		warnp("Failed " #ENDIAN #N "dec");	\
+		print_arr((uint8_t *)&val, N/8);	\
+		goto err0;				\
+	}						\
+} while (0)
+
+/**
+ * TEST_FULL(N, vle, vbe):
+ * Check all endian conversions for $N$ bits, using the values $le$ and $be$.
+ *
+ * Implementation note: Moving the values through $le_v$ and $be_v$ stops the
+ * compiler from generating the answers at compile-time.
+ */
+#define TEST_FULL(N, vle, vbe) do {			\
+	volatile uint ## N ## _t le_v, be_v;		\
+	int system_le = is_system_le();			\
+							\
+	le_v = vle;					\
+	be_v = vbe;					\
+	vle = le_v;					\
+	vbe = be_v;					\
+	if (system_le == 1) {				\
+		TEST_ENDIAN(be, N, vle, vbe, vle, vbe);	\
+		TEST_ENDIAN(be, N, vbe, vle, vbe, vle);	\
+		TEST_ENDIAN(le, N, vle, vle, vle, vle);	\
+		TEST_ENDIAN(le, N, vbe, vbe, vbe, vbe);	\
+	} else {					\
+		TEST_ENDIAN(be, N, vle, vle, vle, vle);	\
```

Comment:
I haven't tested the big-endian code here.  I could theoretically do that in something like QEMU, but at the moment I don't think that's necessary.